src/streams.rs

```diff
@@ -0,0 +1,344 @@
+    ($struct_name: ident) => {
+        pub struct $struct_name {
+            pub stream_type: MediaInfoStream,
+            pub index: usize,
+            pub handler: Rc<RefCell<MediaInfo>>,
+        }
+    }
+}
+
+macro_rules! base_stream_implement {
+    ($struct_name: ident) => {
+        impl BaseStream for $struct_name {
+            fn stream_type(&self) -> MediaInfoStream {
+                self.stream_type
+            }
+
+            fn index(&self) -> usize {
+                self.index
+            }
+
+            fn handler(&self) -> Option<&Rc<RefCell<MediaInfo>>> {
+                Some(&self.handler)
+            }
+        }
+    }
+}
+
+macro_rules! mediainfo_attr {
+    ($meth_name: ident, $attr_name: tt) => (
+        pub fn $meth_name(&self) -> MediaInfoResult<String> {
+            match self.handler() {
+                Some(rc) => rc.borrow_mut().get(self.stream_type(), self.index(), $attr_name, MediaInfoInfo::Text, MediaInfoInfo::Name),
+                None => Err(MediaInfoError::NoDataOpenError)
+            }
+        }
+    )
+}
+
+macro_rules! mediainfo_date {
+    ($meth_name: ident, $attr_name: tt) => (
+        pub fn $meth_name(&self) -> MediaInfoResult<DateTime<UTC>> {
+            match self.handler() {
+                Some(rc) => {
+                    self.result_to_date(rc.borrow_mut().get(self.stream_type(), self.index(), $attr_name, MediaInfoInfo::Text, MediaInfoInfo::Name))
+                },
+                None => Err(MediaInfoError::NoDataOpenError),
+            }
+        }
+    )
+}
+
+macro_rules! mediainfo_i64 {
+    ($meth_name: ident, $attr_name: tt) => (
+        pub fn $meth_name(&self) -> MediaInfoResult<i64> {
+            match self.handler() {
+                Some(rc) => {
+                    self.result_to_i64(rc.borrow_mut().get(self.stream_type(), self.index(), $attr_name, MediaInfoInfo::Text, MediaInfoInfo::Name))
+                },
+                None => Err(MediaInfoError::NoDataOpenError),
+            }
+        }
+    )
+}
+
+macro_rules! mediainfo_duration {
+    ($meth_name: ident, $attr_name: tt) => (
+        pub fn $meth_name(&self) -> MediaInfoResult<Duration> {
+            match self.handler() {
+                Some(rc) => {
+                    self.result_to_duration(rc.borrow_mut().get(self.stream_type(), self.index(), $attr_name, MediaInfoInfo::Text, MediaInfoInfo::Name))
+                },
+                None => Err(MediaInfoError::NoDataOpenError),
+            }
+        }
+    )
+}
+
+pub struct GeneralStream {
+    pub stream_type: MediaInfoStream,
+    pub handler: Option<Rc<RefCell<MediaInfo>>>,
+}
+
+pub trait BaseStream {
+    fn stream_type(&self) -> MediaInfoStream;
+    fn index(&self) -> usize;
+    fn handler(&self) -> Option<&Rc<RefCell<MediaInfo>>>;
+
+    fn result_to_duration(&self, result: MediaInfoResult<String>) -> MediaInfoResult<Duration> {
+        let value = result?;
+        let value = match value.parse::<u64>() {
```

Comment:
What do you think about `result?.parse::<u64>().map_err(|_| MediaInfoError::NonNumericResultError).map(|value| Duration::from_millis(value))`?

If you want to keep the match pattern for consistency you could move the `from_millis` call into the `Ok` block and get rid of the explicit return.